File: Node.js
const fs = require('fs');
const path = require('path');
const { Client, ConsensusMessageSubmitTransaction, Hbar } = require("@hashgraph/sdk");
const Web3 = require("web3");
require('dotenv').config();

// Load JSON configuration file
const configPath = path.join(__dirname, 'config.json');
const config = JSON.parse(fs.readFileSync(configPath, 'utf-8'));

const web3 = new Web3(process.env.INFURA_URL);
const contractAddress = process.env.CONTRACT_ADDRESS;

// Load contract ABI from config file
const contractABI = config.contractABI;

const client = Client.forTestnet().setOperator(process.env.ACCOUNT_ID, process.env.PRIVATE_KEY);

const contract = new web3.eth.Contract(contractABI, contractAddress);

contract.events.StateTransitionLogged({
    fromBlock: 0
}, async (error, event) => {
    if (error) {
        console.error(error);
        return;
    }

    const stateHash = event.returnValues.stateHash;
    const timestamp = event.returnValues.timestamp;

    const tx = new ConsensusMessageSubmitTransaction()
        .setTopicId(process.env.HCS_TOPIC_ID)
        .setMessage(JSON.stringify({ stateHash, timestamp }))
        .setMaxTransactionFee(new Hbar(2));

    try {
        const response = await tx.execute(client);
        console.log("State logged on Hedera with transaction ID:", response.transactionId.toString());
    } catch (err) {
        console.error("Failed to log state on Hedera:", err);
    }
});

File: Node.py
from hedera import (
    Client,
    ConsensusMessageSubmitTransaction,
    Hbar,
    TopicMessageQuery
)
import time
import os
from dotenv import load_dotenv

# Load environment variables from the .env file
load_dotenv()

# Initialize the Hedera client with credentials from environment variables
client = Client.for_testnet()
client.set_operator(os.getenv("ACCOUNT_ID"), os.getenv("PRIVATE_KEY"))

# Topic in HCS, where we are going to send the message
topic_id = os.getenv("HCS_TOPIC_ID")

# Data to write
state_hash = "0x123abc0x123abc0x123abc"  # Replace with your actual state hash
timestamp = int(time.time())  
data = "FIBTATION state data" 

message = f"State Hash: {state_hash}, Timestamp: {timestamp}, Data: {data}"

# Send message to Hedera Consensus Service
transaction = ConsensusMessageSubmitTransaction(
    topic_id=topic_id,
    message=message
).set_max_transaction_fee(Hbar(2))

response = transaction.execute(client)
print(f"State logged on Hedera with transaction ID: {response.transaction_id}")
